pkg/actuators/machine/instances.go

```diff
@@ -440,28 +456,30 @@ func buildDescribeSecurityGroupsTag(tags []machinev1.Tag) *[]ecs.DescribeSecurit
 
 func getVSwitchID(machine runtimeclient.ObjectKey, machineProviderConfig *machinev1.AlibabaCloudMachineProviderConfig, client alibabacloudClient.Client) (string, error) {
 	klog.Infof("validate vswitch in region %s", machineProviderConfig.RegionID)
-	if machineProviderConfig.VSwitch.ID == "" && len(machineProviderConfig.VSwitch.Tags) == 0 {
-		return "", errors.New("no vswitch configuration provided")
-	}
-
-	if machineProviderConfig.VSwitch.ID != "" {
-		return machineProviderConfig.VSwitch.ID, nil
-	}
-
-	if machineProviderConfig.VSwitch.Tags != nil {
+	switch machineProviderConfig.VSwitch.Type {
```

Comment:
Is vSwitch an optional field? Do we need to check for a nil `VSwitch` before we start doing this?

Reply:
it says optional in the api, but that field is not a point to struct but an actual struct. i had _thought_ i tried doing a nil compare here and the compiler barked at me

Reply:
looking through the code again, it seems like although VSwitchID is optional in the API, it would fail if you did not specify one or it could not determine which vswitch to use. i'm thinking we should change the API to make it not optional.

Reply:
these alibaba docs[0], indicate that a vswitch is required when using vpc mode, but that using classic mode does not require one. do we allow machine creation outside of a vpc?

if not, should we make `VpcID`[1] required?

[0] https://www.alibabacloud.com/help/en/doc-detail/63440.html
[1] https://github.com/openshift/api/blob/master/machine/v1/types_alibabaprovider.go#L85

Reply:
asked around and the assumption should be that we require VPC mode. which means that vswitch must be supplied as well.